### DISCUSSION TITLE
PKCS 11 Support

### DISCUSSION THREAD

**Discussion by WilliamBerryiii** [Azure/azure-iot-sdk-csharp#2105] [General]
There is currently a bug/feature request filed against the C SDK for PKCS 11 support and should likely be duplicated for all the other SDK language variants.  Please see this issue for reference: https://github.com/Azure/azure-iot-sdk-c/issues/1246

cc/ @pierreca for python, etc.

**Comment by Christian-e-S** [comment]
Is there any update regarding the PKCS11 support for c#?

As long as there is no support for c#, are there any workarounds to use a HSM chip like the ATECC608B with c# sdk? 

In particular, we search an example for a use case that should be quite common:
 
Platform: Raspberry Pi with Linux
HSM:  ATECC608B
Programming language C# (.net core)

What we would like to achieve, is to use the ATECC608B to authenticate with the IoT Hub without the private key ever leaving the ATECC608B. 

Is there any end-to-end example or guide?

**Reply by mabae** [reply]
> On Linux, .NET Core is using OpenSSL. Using PInvoke it is possible to configure the OpenSSL Engine to use a Hardware Security Module and create an X509Certificate2 object from the context.

Source: https://github.com/Azure-Samples/azure-iot-samples-csharp/tree/master/provisioning/Samples/device/X509Sample
